File: bindings/python/iota_sdk/types/output_id_proof.py
# Copyright 2024 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations
from enum import IntEnum
from typing import Any, Dict, TypeAlias, Union
from dataclasses import dataclass, field
from dataclasses_json import config
from iota_sdk.types.common import HexStr, json, SlotIndex


class OutputCommitmentProofType(IntEnum):
    """OutputCommitmentProof types.

    Attributes:
        HashableNode (0): Denotes a HashableNode.
        LeafHash (1): Denotes a LeafHash.
        ValueHash (2): Denotes a ValueHash.
    """
    HashableNode = 0
    LeafHash = 1
    ValueHash = 2


def deserialize_proof(d: Dict[str, Any]) -> OutputCommitmentProof:
    """
    Takes a dictionary as input and returns an instance of a specific class based on the value of the 'type' key in the dictionary.

    Arguments:
    * `d`: A dictionary that is expected to have a key called 'type' which specifies the type of the returned value.
    """
    proof_type = d['type']
    if proof_type == OutputCommitmentProofType.HashableNode:
        return HashableNode.from_dict(d)
    if proof_type == OutputCommitmentProofType.LeafHash:
        return LeafHash.from_dict(d)
    if proof_type == OutputCommitmentProofType.ValueHash:
        return ValueHash.from_dict(d)
    raise Exception(f'invalid proof type: {proof_type}')


@json
@dataclass
class HashableNode:
    """Contains the hashes of the left and right children of a node in the OutputCommitmentProof tree.
    """
    type: int = field(default_factory=lambda: int(
        OutputCommitmentProofType.HashableNode), init=False)
    l: OutputCommitmentProof = field(metadata=config(
        decoder=deserialize_proof
    ))
    r: OutputCommitmentProof = field(metadata=config(
        decoder=deserialize_proof
    ))


@json
@dataclass
class LeafHash:
    """Contains the hash of a leaf in the OutputCommitmentProof tree.
    """
    type: int = field(default_factory=lambda: int(
        OutputCommitmentProofType.LeafHash), init=False)
    hash: HexStr


@json
@dataclass
class ValueHash:
    """Contains the hash of the value for which the OutputCommitmentProof is being computed.
    """
    type: int = field(default_factory=lambda: int(
        OutputCommitmentProofType.ValueHash), init=False)
    hash: HexStr


@json
@dataclass
class OutputIdProof:
    """The proof of the output identifier.

    Attributes:
        slot: The slot index of the output.
        output_index: The index of the output within the corresponding transaction.
        transaction_commitment: The commitment of the transaction that created the output. Hex-encoded with 0x prefix.
        output_commitment_proof: The proof of the output commitment.
    """
    slot: SlotIndex
    output_index: int
    transaction_commitment: HexStr
    output_commitment_proof: OutputCommitmentProof = field(metadata=config(
        decoder=deserialize_proof
    ))


OutputCommitmentProof: TypeAlias = Union[HashableNode, LeafHash, ValueHash]

File: bindings/nodejs/lib/types/models/api/utxo/output-id-proof.ts
// Copyright 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// Temp solution for not double parsing JSON
import { plainToInstance } from 'class-transformer';
import { SlotIndex } from '../../../block';
import { HexEncodedString } from '../../../utils';

/**
 * OutputCommitmentProof types.
 */
export enum OutputCommitmentProofType {
    /** Denotes a HashableNode. */
    HashableNode = 0,
    /** Denotes a LeafHash. */
    LeafHash = 1,
    /** Denotes a ValueHash. */
    ValueHash = 2,
}

export abstract class OutputCommitmentProof {
    readonly type: OutputCommitmentProofType;

    /**
     * @param type The type of OutputCommitmentProof.
     */
    constructor(type: OutputCommitmentProofType) {
        this.type = type;
    }

    /**
     * Parse an OutputCommitmentProof from a plain JSON object.
     */
    public static parse(data: any): OutputCommitmentProof {
        if (data.type == OutputCommitmentProofType.HashableNode) {
            return plainToInstance(HashableNode, data) as any as HashableNode;
        } else if (data.type == OutputCommitmentProofType.LeafHash) {
            return plainToInstance(LeafHash, data) as any as LeafHash;
        } else if (data.type == OutputCommitmentProofType.ValueHash) {
            return plainToInstance(ValueHash, data) as any as ValueHash;
        }
        throw new Error('Invalid JSON');
    }
}

/**
 * Contains the hashes of the left and right children of a node in the OutputCommitmentProof tree.
 */
export class HashableNode extends OutputCommitmentProof {
    readonly l: OutputCommitmentProof;
    readonly r: OutputCommitmentProof;

    /**
     * @param l Output commitment proof of the left subtree.
     * @param r Output commitment proof of the right subtree.
     */
    constructor(l: OutputCommitmentProof, r: OutputCommitmentProof) {
        super(OutputCommitmentProofType.HashableNode);
        this.l = l;
        this.r = r;
    }
}

/**
 * Contains the hash of a leaf in the OutputCommitmentProof tree.
 */
export class LeafHash extends OutputCommitmentProof {
    readonly hash: HexEncodedString;

    /**
     * @param hash The hash of the leaf.
     */
    constructor(hash: HexEncodedString) {
        super(OutputCommitmentProofType.LeafHash);
        this.hash = hash;
    }
}

/**
 * Contains the hash of the value for which the OutputCommitmentProof is being computed.
 */
export class ValueHash extends OutputCommitmentProof {
    readonly hash: HexEncodedString;

    /**
     * @param hash The hash of the value.
     */
    constructor(hash: HexEncodedString) {
        super(OutputCommitmentProofType.ValueHash);
        this.hash = hash;
    }
}

/**
 * The proof of the output identifier.
 */
export class OutputIdProof {
    readonly slot: SlotIndex;
    readonly outputIndex: number;
    readonly transactionCommitment: HexEncodedString;
    readonly outputCommitmentProof: OutputCommitmentProof;

    /**
     * @param slot The slot index of the output.
     * @param outputIndex The index of the output within the corresponding transaction.
     * @param transactionCommitment The commitment of the transaction that created the output. Hex-encoded with 0x prefix.
     * @param outputCommitmentProof The proof of the output commitment.
     */
    constructor(
        slot: SlotIndex,
        outputIndex: number,
        transactionCommitment: HexEncodedString,
        outputCommitmentProof: OutputCommitmentProof,
    ) {
        this.slot = slot;
        this.outputIndex = outputIndex;
        this.transactionCommitment = transactionCommitment;
        this.outputCommitmentProof = outputCommitmentProof;
    }
}
